src/Utils/XPath.php

```diff
@@ -0,0 +1,32 @@
+<?php
+
+namespace RobRichards\XMLSecLibs\Utils;
+
+class XPath
+{
+    const ALPHANUMERIC = 0;
+    const NUMERIC = 1;
+    const LETTERS = 2;
+    const EXTENDED_ALPHANUMERIC = 3;
+
```

Comment:
This misses some of the non-alpha/numeric values that are allowed:
https://www.w3.org/TR/REC-xml/#id

Reply:
Hi Rob,

Do you have any suggestions for non-alphanumeric characters that should be allowed? Or you mean every character allowed by the recommendation (1)?

My initial fix for this was to filter single and double quotes out, which would be much simpler (blacklist). However, the current best practice recommends whitelisting instead of blacklisting, and this blacklisting approach wouldn't even work when you are building an attribute's name based (possibly) on user input, like the one coming from `$idKey`.

In the end, I don't think most of the characters allowed by the recommended are used or even reasonable. At least, I don't find reasonable having an ID (or an attribute name, for that matter) like `⁰‿⁰`, even if that's perfectly legal.


(1) `Name` production:

NameStartChar | ::= | ":" \| [A-Z] \| "_" \| [a-z] \| [#xC0-#xD6] \| [#xD8-#xF6] \| [#xF8-#x2FF] \| [#x370-#x37D] \| [#x37F-#x1FFF] \| [#x200C-#x200D] \| [#x2070-#x218F] \| [#x2C00-#x2FEF] \| [#x3001-#xD7FF] \| [#xF900-#xFDCF] \| [#xFDF0-#xFFFD] \| [#x10000-#xEFFFF]
-- | -- | --
NameChar | ::= | NameStartChar \| "-" \| "." \| [0-9] \| #xB7 \| [#x0300-#x036F] \| [#x203F-#x2040]
Name | ::= | NameStartChar (NameChar)*

Reply:
I would at least add :_#-.
My concern is that if someone is using some software to produce the document that it may randomly generate values based on the ruleset which would cause the document to never be able to be verified by the library.

Reply:
I see your point. Let me try something else, differentiating between attribute names and values.